### ISSUE TITLE
QName literals

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#747]
It's quite common to want to write a constant QName; I found myself doing this a lot, for example, in examples and test cases for the `elements-to-maps()` function. It's clumsy having to call `xs:QName()` or `fn:QName()` or `parse-QName()` for this purpose. It's particularly clumsy with map constructors where you want to write many QNames.

I propose we introduce the syntax `Q"prefix:local"`.

The quotes can be either single or double; the prefix is optional. If there is no prefix, the result is a no-namespace QName.

The prefix (if present) must be bound to a namespace in the static context.

Character and entity references are not allowed.

Note that `Q{uri}local` is a NameTest, not a QName literal.

**Comment by dnovatchev** [comment]
Just to understand better:

With this proposal, will anyone still have to write:

```xq
let $put := function($map as map(*), $key as xs:anyAtomicType, $value as item()*) as map(*)
             { Q{http://www.w3.org/2005/xpath-functions/map}put($map, $key, $value) },
             
    $array-head := function($array as array(*)) as item()*
             { Q{http://www.w3.org/2005/xpath-functions/array}head($array)},

    $array-tail := function($array as array(*)) as array(*)
             { Q{http://www.w3.org/2005/xpath-functions/array}tail($array)},

    $array-size := function($array as array(*)) as xs:integer
             { Q{http://www.w3.org/2005/xpath-functions/array}size($array)},

    $array-join := function($arrays as array(*)*) as array(*)
             { Q{http://www.w3.org/2005/xpath-functions/array}join($arrays)}
  return
     (: Whatever :)
```

**Comment by ChristianGruen** [comment]
@michaelhkay Maybe we could also allow `Q"{uri}local"`.

> With this proposal, will anyone still have to write: […]

@dnovatchev Right.

**Comment by dnovatchev** [comment]
> @michaelhkay Maybe we could also allow `Q"{uri}local"`.
> 
> > With this proposal, will anyone still have to write: […]
> 
> @dnovatchev Right.

You mean, one still will have to write the below? I would value much more a  change that significantly simplifies this :

```xq
let $put := function($map as map(*), $key as xs:anyAtomicType, $value as item()*) as map(*)
             { Q{http://www.w3.org/2005/xpath-functions/map}put($map, $key, $value) },
             
    $array-head := function($array as array(*)) as item()*
             { Q{http://www.w3.org/2005/xpath-functions/array}head($array)},

    $array-tail := function($array as array(*)) as array(*)
             { Q{http://www.w3.org/2005/xpath-functions/array}tail($array)},

    $array-size := function($array as array(*)) as xs:integer
             { Q{http://www.w3.org/2005/xpath-functions/array}size($array)},

    $array-join := function($arrays as array(*)*) as array(*)
             { Q{http://www.w3.org/2005/xpath-functions/array}join($arrays)}
  return
     (: Whatever :)
```

**Comment by ChristianGruen** [comment]
> I would value much more a change that significantly simplifies this :
> { Q{http://www.w3.org/2005/xpath-functions/map}put($map, $key, $value) },

@dnovatchev I can’t imagine that this can be achieved by introducing a syntax for QName literals. Maybe a separate issue would be a better place to propose how you think that the existing syntax could be simplified?

**Comment by michaelhkay** [comment]
The syntax for QName literals doesn't affect how you write function names, variable names, or (node) name tests. It affects cases where you need to construct a QName value when calling a function that requires one, for example

```
error(Q"err:XPTY0004")
serialize($data, map{Q"saxon:double-space", Q"xhtml:p"})
```

**Comment by dnovatchev** [comment]
> The syntax for QName literals doesn't affect how you write function names, variable names, or (node) name tests. It affects cases where you need to construct a QName value when calling a function that requires one, for example
> 
> ```
> error(Q"err:XPTY0004")
> serialize($data, map{Q"saxon:double-space", Q"xhtml:p"})
> ```

@michaelhkay I believe we already have (had) a proposal for specifying namespaces in XPath, right?

**Comment by ChristianGruen** [comment]
We could as well try to promote `xs:string` to `xs:QName` (related: #480). Looking at the examples…

```xquery
error(Q"err:XPTY0004")
serialize($data, map{Q"saxon:double-space", Q"xhtml:p"})
```

…I assume it would only work for the first case, though.

**Comment by michaelhkay** [comment]
Promoting string to QName would be very problematic because it's not clear what the static context for the promotion should be, and because the string might be constructed at run-time. This proposal is all about xs:QName literals.

**Comment by ndw** [comment]
It feels like we keep adding more and more ad-hoc syntax for things. I don't dispute that `xs:QName("xhtml:p")` is a bit longer, but is it *so* much longer that it justifies changes to the parser to match `Q"xhtml:p"`?

It feels like if we want to go this way, it would be nice to find a way to harmonize EQName (name tests) at the same time.

**Comment by ChristianGruen** [comment]
I tend to agree with Norm that we shouldn’t overdo it with even more new constructors. My personal experience is that most non-expert users would like not to care about QNames at all, and would rather be happy if they could just write things like `error('range', 'Out of range!')` (i.e., use simple strings instead of QNames, which could still be dynamically converted to QNames).

**Comment by ndw** [comment]
FWIW, XProc attempts to do some of this: https://spec.xproc.org/3.0/xproc/#qname-handling

(Whether it got that exactly right, and whether similar sorts of rules would apply in **this** context is not something I've considered carefully; just passing along a pointer.)

**Comment by michaelhkay** [comment]
So, it seems there's little enthusiasm for a QName literal syntax, but some enthusiasm for extensions to the coercion rules for the case where the required type is xs:QName. Specifically, add a rule that if a string or untypedAtomic value is supplied where the required type is xs:QName, then the supplied value is effectively wrapped in a call of `fn:parse-QName()`.

The tricky part here is that this creates a dependency on the namespace context. Which namespace context, precisely? For a function call, the namespace context of the function call (not the declaration) -- or rather, the namespace context of the specific argument in the function call? But the coercion rules are used for things other than function calls, so there's potential for some complexity here. The real danger comes when it's a dynamic function call, so you don't know statically that you're introducing a dependency on the namespace context, which means you don't know statically that the namespace context needs to be retained for use at run time.

**Comment by ChristianGruen** [comment]
> The tricky part here is that this creates a dependency on the namespace context. Which namespace context, precisely? For a function call, the namespace context of the function call (not the declaration) -- or rather, the namespace context of the specific argument in the function call?

The latter version would probably be cleaner to realize. Which namespace context would be used by `Q"…"`, i.e., what would be the result of the following expression?

```xquery
data(<_ xmlns='URI'>{ namespace-uri-from-QName(Q"_") }</_>)
```

Since we already have the `Q{URI}name` syntax, the proposed literal syntax would be somewhat familiar, so… I’m still wavering.

**Comment by ndw** [comment]
The group agreed to close this issue without action at [meeting 054](https://qt4cg.org/meeting/minutes/2023/11-14.html)

**Comment by ChristianGruen** [comment]
Redundant to #1661 (even though we haven’t decided what to do).

**Comment by ndw** [comment]
At [meeting 107](https://qt4cg.org/meeting/minutes/2025/01-28.html), the CG agreed to close this issue without further action.
